typing/signature_group.ml

```diff
@@ -0,0 +1,116 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*  Florian Angeletti, projet Cambium, Inria Paris                        *)
+(*                                                                        *)
+(*   Copyright 2021 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Fold on a signature by syntactic group of items *)
+
+(** Classes and class types generate ghosts signature items, we group them
+    together before printing *)
+type sig_item =
+  {
+    src: Types.signature_item;
+    post_ghosts: Types.signature_item list
+    (** ghost classes types are post-declared *);
+  }
+let flatten x = x.src :: x.post_ghosts
+
+type core_rec_group =
+  | Not_rec of sig_item
+  | Rec_group of sig_item list
+
+let rec_items = function
+  | Not_rec x -> [x]
+  | Rec_group x -> x
+
+(** Private row types are manifested as a sequence of definitions
+    preceding a recursive group, we collect them and separate them from the
+    syntatic recursive group. *)
+type rec_group =
+  { pre_ghosts: Types.signature_item list; group:core_rec_group }
+
+let take n seq =
+  let rec aux l rem n seq =
+    if n = 0 then List.rev l, rem, seq else
+      match seq () with
+      | Seq.Nil -> assert false
+      | Seq.Cons((x,rem),next) ->
+          aux (x::l) rem (n-1) next
+  in
+  aux [] [] n seq
+
+let rec partial_lists l () = match l with
+  | [] -> Seq.Nil
+  | a :: q -> Seq.Cons((a,q), partial_lists q)
```

Comment:
I don't like this `partial_lists` business, it feels too complex (and not so composable). Can you explain when the apparently-redundant `q` data is important/useful?

Reply:
It is used to edit only one item in a signature in the case of a `with` constraint without traversing the whole signature.
The other option would be to reconstruct the remaining signature from the structured sequence.
Or I could be also convinced to switch to a zipper.